### PR TITLE
Refactor configuration to support SetVariables

### PR DESCRIPTION
**Pull request by Bholtland**
Configuration is now a class with common methods for OCPP 2.0.1 and 1.6.
The underlying data structure for those versions is different. For 2.0.1 this is not just a flat array of objects holding just the key values, but is closely modeled after the OCPP spec. (That way we can later do more advanced things like changing the non-Actual variable attributes).

Also added a handler for `SetVariables` message

### CODE REVIEW

File: src/screens/Dashboard/index.js [hunk 1]
```diff
@@ -218,32 +215,29 @@ export default class Home extends React.Component {
...
 
+                    this.setState({
+                      configuration: newConfiguration,
+                    });
+                    chargeStation.changeConfiguration(newConfiguration);
+                  } else {
+                    configuration.updateVariablesFromSimpleConfigurationMap(
+                      config
+                    );
+                  }
```
Comment (Bholtland) [inline src/screens/Dashboard/index.js:235]: This still requires some improvement. Right now you first have to change the OCPP version, then 'save', then go back to the settings and change all the other configurations as we don't port 1.6 variables to 2.0.1 and vice versa.

Comment (Bholtland) [reply src/screens/Dashboard/index.js:235]: Will think about this a bit more.

File: src/lib/ChargeStation/eventHandlers/ocpp-16/handle-change-configuration.js [hunk 2]
```diff
@@ -3,8 +3,15 @@ export default async function handleChangeConfiguration({
...
-  chargepoint.configuration[key] = value;
+  const { key } = callMessageBody;
+
+  chargepoint.configuration.setVariable(key, callMessageBody);
+
+  if (chargepoint.configuration[key]) {
+    chargepoint.configuration[key].value = value;
+  } else {
+    chargepoint.configuration[key] = callMessageBody;
+  }
```
Comment (lucasvmiguel) [inline src/lib/ChargeStation/eventHandlers/ocpp-16/handle-change-configuration.js:14]: Why don't we place the following code into `chargepoint.configuration.setVariable(key, callMessageBody)`?
```js
  if (chargepoint.configuration[key]) {
    chargepoint.configuration[key].value = value;
  } else {
    chargepoint.configuration[key] = callMessageBody;
  }
```

File: src/lib/settings.js [hunk 3]
```diff
@@ -526,26 +529,156 @@ export function getDefaultSession() {
   return result;
 }
 
-function buildConfigurationMap201() {
-  const query = getDocumentQuery();
-  const list = configurationList201;
-  const result = {};
+class VariableConfiguration201 {
```
Comment (lucasvmiguel) [inline src/lib/settings.js:532]: I would inherit a class like `VariableConfiguration` in both `VariableConfiguration201` and `VariableConfiguration16` just to be clear that those classes share common methods (and must implement some of them)

File: src/lib/settings.js [hunk 4]
```diff
@@ -526,26 +529,156 @@ export function getDefaultSession() {
...
-    if (query.get(key)) {
-      result[key] = query.get(key);
-      continue;
+  getMeterValueSampleInterval() {
+    const defaultInterval = 60;
+    const value = this.getVariableActualValue('MeterValueSampleInterval');
+    return value || defaultInterval;
+  }
+
+  variablesToSimpleConfigurationMap() {
```
Comment (lucasvmiguel) [inline src/lib/settings.js:551]: I'm seeing this `simple` in a few different places. I wonder if we should call something more straightforward to make it easier to understand what simple means. (eg: `variablesToKeyValueMap`)

name is just a suggestion, you can choose another one.

**Review by lucasvmiguel — APPROVED** [review]

File: src/lib/settings.js (hunk 4 above)
Comment (Bholtland) [reply src/lib/settings.js:551]: Good one, was struggling with this as well. The one you suggested might be nice actually.

**Pull request merged**